rmk/src/keyboard.rs

```diff
@@ -411,105 +335,32 @@ impl<'a, M: MatrixTrait, const ROW: usize, const COL: usize, const NUM_LAYER: us
-    ) {
-        if key_state.is_pressing() {
-            self.process_key_action_normal(action, key_state, sender)
-                .await;
-
-            // Wait 10ms, then send release
-            Timer::after_millis(10).await;
-            // Manually trigger send report
-            self.send_keyboard_report(sender).await;
-            self.send_media_report(sender).await;
-            self.send_system_control_report(sender).await;
-            self.send_mouse_report(sender).await;
-
-            key_state.pressed = false;
-            self.process_key_action_normal(action, key_state, sender)
-                .await;
-        }
-    }
-
-    /// Process tap/hold action.
-    /// There are several cases:
-    ///
-    /// 1. `key_state.changed` is true, and `key_state.pressed` is false,
-    ///     which means the key is released. Then the duration time should be checked
-    ///
-    /// 2. `key_state.changed` is false, and `key_state.pressed` is true,
-    ///     which means that the key is held. The duration time should to be checked.
-    ///     
-    /// TODO: make tap/hold threshold customizable
-    async fn process_key_action_tap_hold(
-        &mut self,
-        tap_action: Action,
-        hold_action: Action,
-        row: usize,
-        col: usize,
-        mut key_state: KeyState,
-        sender: &Sender<'a, CriticalSectionRawMutex, KeyboardReportMessage, 8>,
-    ) {
-        if key_state.is_releasing() {
-            // Case 1, the key is released
-            if let Some(s) = key_state.hold_start {
-                let d = s.elapsed().as_millis();
-                if d < 200 {
-                    // Released: tap
-                    key_state.pressed = true;
-                    debug!("Release tap hold, got TAP: {}, {}", tap_action, key_state);
-                    self.process_key_action_tap(tap_action, key_state, sender)
-                        .await;
-                } else {
-                    // Released: hold action
-                    debug!("Release tap hold, got HOLD: {}, {}", hold_action, key_state);
-                    self.process_key_action_normal(hold_action, key_state, sender)
-                        .await;
-                }
-                // Reset timer after release
-                self.matrix.update_key_state(row, col, |ks| {
-                    ks.clear_timer();
-                });
-            }
-        } else if key_state.pressed && !key_state.changed {
-            // Case 2, the key is held
-            if let Some(s) = key_state.hold_start {
-                let d = s.elapsed().as_millis();
-                if d > 200 {
-                    self.process_key_action_normal(hold_action, key_state, sender)
-                        .await;
-                }
-            }
+    async fn process_key_action_tap(&mut self, action: Action, mut pressed: bool) {
+        if pressed {
+            self.process_key_action_normal(action, pressed).await;
```

Comment:
A proper delay(> 1ms) should be set between two reports of tap

Reply:
Any delay adds to the additional 0.1ms delay you set before. By removing all the additional delays, I made scanner more responsive at guaranteed 1k per second.

This is the speed write to channel, no delay needed. Only the communication task, write to USB or BLE need delay after each report.

I think the channel should be a lot longer than 8. To avoid delay caused by channel full. Especially the Macro playback, need very long channel. Or move the macro playback to communication thread, so we can accommodate long macro without a long channel.